### PR TITLE
Make OSSL_TIME a structure

### PR DESCRIPTION
**Pull request by paulidale**
This prevents misuses creeping in that would make converting later difficult.

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by paulidale** [comment]
The _weak_ attribute on the definition of `ossl_time_now()` needs to go.  Without it, a link time error is being raised build the event queue test:
> /usr/bin/ld: libssl.a(libssl-lib-time.o): in function `ossl_time_now':
/home/pauli/work/ossl/../openssl/ssl/time.c:18: multiple definition of `ossl_time_now'; test/event_queue_test-bin-event_queue_test.o:/home/pauli/work/ossl/../openssl/test/event_queue_test.c:17: first defined here

This didn't happen when the function returned an `int` which is pretty confusing.

**Comment by t8m** [comment]
Why would we ever want the time to be anything else than 64bit signed or unsigned integer?

**Comment by hlandau** [comment]
The only remotely likely change in the future would be to a `(s, ns)` tuple to match the precision and range of `struct timespec`. However the current range and precision is good enough for our purposes so this is not terribly likely.

What this does do is ensure people use the right functions to do arithmetic on `OSSL_TIME` values, as if we don't make it a structure slipups where people just do integer arithmetic on them directly are  bound to creep in. So this seems like a good idea to me.

**Comment by paulidale** [comment]
@hlandau got it spot on.  If we decide to change in the future, all problems will need to be addressed then.  By making it a structure now, there won't be any (hopefully).

The x86/64 code generated seems to be the same, at least for the couple of bits I've looked at.  This won't hold true for all platforms I suspect.  It will depend on the ABI's handling of short structures.

**Comment by paulidale** [comment]
Seems that the _weak_ attribute wasn't sufficient by itself on some platforms.
This one's got be confused, at least for the moment.

**Comment by mattcaswell** [comment]
CI failures are relevant

**Review by levitte — APPROVED** [review]

File: ssl/time.c [hunk 1]
```diff
@@ -11,8 +11,13 @@
 #include <openssl/err.h>
 #include "internal/time.h"
 
+OSSL_TIME OSSL_TIME_INFINITY = { ~(uint64_t)0 };
+OSSL_TIME OSSL_TIME_IMMEDIATE = { 0 };
```
Comment (levitte) [inline ssl/time.c:15]: Put these in a separate source file.  That'll get rid of the linking problem that this PR currently experiences

File: doc/internal/man3/OSSL_TIME.pod [hunk 2]
```diff
@@ -2,19 +2,23 @@
...
 
- typedef uint64_t OSSL_TIME;
+ typedef struct OSSL_TIME;
 
- #define OSSL_TIME_SECOND
- #define OSSL_TIME_INFINITY
+ #define OSSL_TIME_SECOND   /* Ticks per second */
+
+ OSSL_TIME OSSL_TIME_INFINITY;
+ OSSL_TIME OSSL_TIME_IMMEDIATE;
```
Comment (levitte) [inline doc/internal/man3/OSSL_TIME.pod:19]: This re-introduces global variables, which find-doc-nits currently doesn't understand.  Hmm...

File: include/internal/time.h [hunk 3]
```diff
@@ -15,21 +15,25 @@
...
-typedef uint64_t OSSL_TIME;
+typedef struct {
+    uint64_t t;
+} OSSL_TIME;
+
+/* The precision of times allows this many values per second */
+# define OSSL_TIME_SECOND 1000000000
+
+/* Value representing the most distant future time */
+extern OSSL_TIME OSSL_TIME_INFINITY;
```
Comment (levitte) [inline include/internal/time.h:33]: Shouldn't this be const?

File: include/internal/time.h [hunk 4]
```diff
@@ -15,21 +15,25 @@
...
+
+/* Value that's guaranteed to be now or before */
+extern OSSL_TIME OSSL_TIME_IMMEDIATE;
```
Comment (levitte) [inline include/internal/time.h:36]: ... and this?

**Review by levitte — CHANGES_REQUESTED** [review]

File: doc/internal/man3/OSSL_TIME.pod (hunk 2 above)
Comment (t8m) [reply doc/internal/man3/OSSL_TIME.pod:19]: They should be also const/

File: ssl/time.c (hunk 1 above)
Comment (t8m) [reply ssl/time.c:15]: They should be const

Comment (paulidale) [reply ssl/time.c]: Makes sense @levitte, thanks.  No idea why I didn't realise :disappointed:

File: doc/internal/man3/OSSL_TIME.pod (hunk 2 above)
Comment (paulidale) [reply doc/internal/man3/OSSL_TIME.pod]: I've made them inline functions.  This avoids the const question and the globals.

**Comment by paulidale** [comment]
CI failure is not relevant.

**Review by levitte — APPROVED** [review]

**Comment by paulidale** [comment]
Thanks @levitte :grin:

**Review by hlandau — APPROVED** [review]

**Comment by t8m** [comment]
This now conflicts and needs to be rebased.

**Comment by paulidale** [comment]
integer / time contamination fixed.
Could someone please merge this if it gets reapproved?

File: include/internal/time.h [hunk 5]
```diff
@@ -15,53 +15,70 @@
...
+ * The time datum is Unix's 1970 and at nanosecond precision, this gives
+ * a range of 584 years roughly.
+ */
+typedef struct {
+    uint64_t t;     /* Ticks since the epoch */
+} OSSL_TIME;
+
 /* The precision of times allows this many values per second */
-# define OSSL_TIME_SECOND ((uint64_t)1000000000)
+# define OSSL_TIME_SECOND 1000000000
```
Comment (hlandau) [inline include/internal/time.h:30]: Why this change?

**Review by hlandau — CHANGES_REQUESTED** [review]

File: include/internal/time.h (hunk 5 above)
Comment (paulidale) [reply include/internal/time.h:30]: Didn't think it was necessary....

Comment (paulidale) [reply include/internal/time.h]: Undone.

**Comment by hlandau** [comment]
Just a doc nit waiting to be resolved.

**Comment by tmshort** [comment]
Noting that the `SSL_SESSION` timeout code (which adds `timeout` to `time`) probably ought to be updated to use `OSSL_TIME`; however, there seems to be no API to convert `time_t` or `struct timeval` to an OSSL_TIME?

**Comment by paulidale** [comment]
> Noting that the `SSL_SESSION` timeout code (which adds `timeout` to `time`) probably ought to be updated to use `OSSL_TIME`; however, there seems to be no API to convert `time_t` or `struct timeval` to an OSSL_TIME?

Yes, it should be converted in due course.  Yes, there isn't such an API yet.

**Comment by paulidale** [comment]
ping for review

File: test/event_queue_test.c [hunk 6]
```diff
@@ -31,38 +31,46 @@ static int event_test(void)
...
+            || !TEST_ptr(e4 = ossl_event_queue_add_new(q, 2, 5,
+                                                       ossl_ticks2time(1150),
+                                                       "ctx 2",
                                                        PAYLOAD("data")))
 
             /* Verify some event details */
             || !TEST_uint_eq(ossl_event_get_type(e1), 1)
             || !TEST_uint_eq(ossl_event_get_priority(e1), 10)
-            || !TEST_uint64_t_eq(ossl_event_get_when(e1), 1100)
+            || !TEST_uint64_t_eq(ossl_event_get_when(e1).t, 1100)
```
Comment (mattcaswell) [inline test/event_queue_test.c:53]: Shouldn't this be using `ossl_time2ticks`? Same for other similar instances below.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Review by mattcaswell — APPROVED** [review]

**Review by hlandau — APPROVED** [review]

**Comment by mattcaswell** [comment]
Pushed to master. Thanks.
